cairo-libs/packages/protobuf/src/types/message.cairo

```diff
@@ -72,85 +78,122 @@ pub impl DecodeContextImpl of DecodeContextTrait {
         DecodeContext { buffer, index: 0, limits: array![] }
     }
 
-    fn init_branch(ref self: DecodeContext, length: usize) {
+    fn init_branch(ref self: DecodeContext, length: usize) -> bool {
         let limit = self.index + length;
-        assert(limit <= self.buffer.len(), 'bigger limit for branch');
+        if limit > self.buffer.len() {
+            return false;
+        }
         self.limits.append(self.index + length);
+        true
     }
 
     fn can_read_branch(ref self: DecodeContext) -> bool {
         @self.index < self.limits[self.limits.len() - 1]
     }
 
+    fn decode_raw<T, +Drop<T>, +Default<T>, +ProtoMessage<T>>(
+        ref self: DecodeContext,
+    ) -> Option<T> {
+        ProtoMessage::<T>::decode_raw(ref self)
+    }
+
     fn decode_field<T, +ProtoMessage<T>, +Drop<T>, +Default<T>>(
-        ref self: DecodeContext, field_number: u8, ref value: T,
-    ) {
+        ref self: DecodeContext, field_number: u8, ref field: T,
+    ) -> bool {
```

Comment:
do you think, it's be good to return `Option<T>` (with `+Default<T>` bound) here? in that way we can avoid the `if` conditions. I liked that before the decode method was clean.

with my suggestion, `Height` decoding will be

```cairo
    fn decode_raw(ref context: DecodeContext) -> Option<Height> {
        let revision_number = context.decode_field(1)?;
        let revision_height = context.decode_field(2)?;
        
        Ok(Height{
            revision_number,
            revision_height,
        })
    }
```

Reply:
If `?` operator is not available, you can use `unwrap()`. but if `?` is available, maybe we can simplify many `if` branches introduced by _boolean-as-success-status_.

Reply:
`unwrap()` wasn’t an option since I wanted to return early on each decoding failure. Not sure why I thought we don't have `?` in Cairo. Things just got simpler. Appreciate you pointing that out! [b0810d0](https://github.com/informalsystems/ibc-starknet/pull/281/commits/b0810d0c93a4c77a74b66b69e98d7176ce0ba992)